# Purpose:
- Dev response analysis from lims
- Eventually going to be implemented in CodeOcean

In [2]:
from pathlib import Path
import pandas as pd
import numpy as np
from pprint import pprint

In [3]:
session_dir = Path(r'\\allen\programs\mindscope\production\learning\prod1\specimen_1341432280\ophys_session_1365173010')
osid = int(session_dir.name.split('_')[-1])
pkl_fn = session_dir / f'{osid}.pkl'
df = pd.read_pickle(pkl_fn)


In [4]:
df.keys()

dict_keys(['session_uuid', 'trigger_delay_sec', 'syncpulselines', 'di', 'mouse_id', 'nidaq_tasks', 'post_blank_sec', 'comp_id', 'monitor_brightness', 'stopdatetime', 'user_id', 'monitor', 'script', 'pre_blank_sec', 'rig_id', 'startdatetime', 'platform', 'window', 'wwidth', 'params', 'stop_time', 'scripttext', 'droppedframes', 'config', 'do', 'syncpulse', 'monitor_contrast', 'start_time', 'escape_pressed', 'wheight', 'syncsqrsize', 'syncpulseport', 'vsynccount', 'stage', 'fps', 'intervalsms', 'sweepstim_text', 'stimuli', 'config_path', 'movie_output', 'sync_sqr_loc', 'items', 'showmouse', 'unpickleable', 'total_frames', 'primary_stimulus', 'sync_config'])

In [18]:
df['stimuli'][1]['display_sequence']

array([[ 840, 1680],
       [2880, 3720]])

In [12]:
df['pre_blank_sec']

20

In [14]:
df['stimuli'][1]['sweep_frames']

[(0, 119),
 (180, 299),
 (360, 479),
 (540, 659),
 (720, 839),
 (900, 1019),
 (1080, 1199),
 (1260, 1379),
 (1440, 1559),
 (1620, 1739),
 (1800, 1919),
 (1980, 2099),
 (2160, 2279),
 (2340, 2459),
 (2520, 2639),
 (2700, 2819),
 (2880, 2999),
 (3060, 3179),
 (3240, 3359),
 (3420, 3539),
 (3600, 3719),
 (3780, 3899),
 (3960, 4079),
 (4140, 4259),
 (4320, 4439),
 (4500, 4619),
 (4680, 4799),
 (4860, 4979),
 (5040, 5159),
 (5220, 5339),
 (5400, 5519),
 (5580, 5699),
 (5760, 5879),
 (5940, 6059),
 (6120, 6239),
 (6300, 6419),
 (6480, 6599),
 (6660, 6779),
 (6840, 6959),
 (7020, 7139),
 (7200, 7319),
 (7380, 7499),
 (7560, 7679),
 (7740, 7859),
 (7920, 8039),
 (8100, 8219),
 (8280, 8399),
 (8460, 8579),
 (8640, 8759),
 (8820, 8939),
 (9000, 9119),
 (9180, 9299),
 (9360, 9479),
 (9540, 9659),
 (9720, 9839),
 (9900, 10019),
 (10080, 10199),
 (10260, 10379),
 (10440, 10559),
 (10620, 10739),
 (10800, 10919),
 (10980, 11099),
 (11160, 11279),
 (11340, 11459),
 (11520, 11639),
 (11700, 11819),
 (

In [6]:
df

{'session_uuid': '66269dd3-2b3b-4395-b5a5-b0c67dd5cc15',
 'trigger_delay_sec': 5.0,
 'syncpulselines': [5, 6],
 'di': None,
 'mouse_id': '726433',
 'nidaq_tasks': {},
 'post_blank_sec': 2,
 'comp_id': 'MESO.1-Stim',
 'monitor_brightness': 37,
 'stopdatetime': datetime.datetime(2024, 5, 9, 9, 24, 41, 524000),
 'user_id': 'vivian.ha',
 'monitor': {'sizepix': (1920, 1200),
  'name': 'Gamma1.Luminance50',
  'gammagrid': [[0.0,
    1.0,
    2.248464822769165,
    7256.35205078125,
    nan,
    6.2604241371154785],
   [0.0, 1.0, 2.257481098175049, 2415.508544921875, nan, 1.5005038976669312],
   [0.0, 1.0, 2.28798770904541, 2434.026611328125, nan, 2.215834856033325],
   [0.0, 1.0, 2.199329137802124, 2406.731689453125, nan, 2.5777041912078857]],
  'distancecm': 15,
  'widthcm': 52,
  'calibrationdate': '1547226229.0',
  'gamma': array([2.257481 , 2.2879877, 2.1993291], dtype=float32)},
 'script': 'C:/ProgramData/AIBS_MPE/camstim/logs/240509082200\\STAGE_1_240509082200.py',
 'pre_blank_sec': 20

In [17]:
df['stimuli'][1]

{'save_sweep_table': True,
 'shuffle': True,
 'frame_list': array([-1, -1, -1, ..., -1, -1, -1]),
 'blank_length': 1,
 'stim_path': '\\\\allen\\programs\\mindscope\\workgroups\\omfish\\Visual Stimuli\\drifting_gratings_TF.stim',
 'stim': "GratingStim(autoDraw=False, autoLog=True, color=array([1., 1., 1.]), colorSpace='rgb', contrast=0.8, depth=0, interpolate=False, mask='None', maskParams=None, name='unnamed GratingStim', opacity=1.0, ori=0.0, phase=array([29751.86666667, 29751.86666667]), pos=array([0., 0.]), rgbPedestal=array([0., 0., 0.]), sf=array([0.04, 0.04]), size=array([250., 250.]), tex='sin', texRes=256, units='deg', win=Window(...))",
 'stim_text': 'stimulus = Stimulus(visual.GratingStim(window,\n                    pos=(0, 0),\n                    units=\'deg\',\n                    size=(250, 250),\n                    mask="None",\n                    texRes=256,\n                    sf=0.1,\n                    ),\n    sweep_params={\n               \'Contrast\': ([0.8],

In [7]:
import h5py
import warnings
# Loading sync file
h5_fn = '1365173010_20240509T082142.h5'
hdf_file_path = session_dir / h5_fn
hdf = h5py.File(hdf_file_path, "r")

stim_info = pd.read_pickle(pkl_fn)
 
print(hdf.keys())
data = hdf['data'][:]
sync_sig = data[:, -1]
time_stamps = data[:, 0]
 
 
# Read meta data
meta = eval(str(np.array(hdf['meta'])).replace('b"','').replace('"',''))
line_labels = meta['line_labels']
sample_freq = meta['ni_daq']['sample_rate']
 
# !! Since I cannot read meta right now !!
# !! this should replace with a proper code that reads these variables from meta!!
 
# line_labels = ['vsync_2p', '', 'vsync_stim', '', 'stim_photodiode', 'stim_running', '', '', 'beh_frame_received', 'eye_frame_received', 'face_frame_received', '', '', '', '', '', '', 'stim_running_opto', 'stim_trial_opto', '', '', 'beh_cam_frame_readout', 'face_cam_frame_readout', '', '', 'eye_cam_frame_readout', '', 'beh_cam_exposing', 'face_cam_exposing', 'eye_cam_exposing', '', 'lick_sensor']
# sample_freq = 100000
 
# Finding rising and falling edges
falling_edges = {}
rising_edges = {}
 
for bit, label in enumerate(line_labels):
    if not label:
        label = 'sig_' + str(bit)
 
    bit_array = np.bitwise_and(sync_sig, 2 ** bit).astype(bool).astype(np.uint8)
    bit_changes = np.ediff1d(bit_array, to_begin=0)
    falling_edges[label] = time_stamps[np.where(bit_changes == 255)]/sample_freq
    rising_edges[label] = time_stamps[np.where(bit_changes == 1)]/sample_freq
 
#Test and correct for photodiode transition errors (look for the last pulse-train pattern, 3 short pulses)
 
ptd_rise_diff = np.ediff1d(rising_edges['stim_photodiode'])
short = np.where(np.logical_and(ptd_rise_diff > 0.1, ptd_rise_diff < 0.3))[0]
medium = np.where(np.logical_and(ptd_rise_diff > 0.5, ptd_rise_diff < 1.5))[0]
 
# find the time differences betwen photodiode rising edges:
# between 0.1 and 0.3 s -> short
# between 0.5 and 1.5 s -> medium
# find tha largest medium that there is a short comes just before or one before that
# ptd_start = that medium + 1 otherwise ptd_start = 4
ptd_start = 3
for i in medium:
    if set(range(i - 2, i)) <= set(short):
        ptd_start = i + 1
 
if ptd_start > 3:
    print('ptd_start: ' + str(ptd_start))
    print("Photodiode events before stimulus start.  Deleted.")
 
 
if rising_edges['stim_photodiode'].max() <= falling_edges['vsync_stim'].max():
    print('photodiode ends before stim_vsync already.')
    ptd_end = len(ptd_rise_diff)
else:
    print('truncating photodiode to end before stim_vsync.')
    ptd_end = np.where(rising_edges['stim_photodiode'] > falling_edges['vsync_stim'].max())[0][0] - 1
print('ptd_end: ' +str(ptd_end)+ ' max photodiode ' + str(rising_edges['stim_photodiode'].max())+' max stim '+ str(falling_edges['vsync_stim'].max()))
ptd_errors = []
error_frames = np.where(ptd_rise_diff[ptd_start:ptd_end] < 1.8)[0] + ptd_start
if len(error_frames)>0:
    print("Photodiode error detected. Number of frames:", len(error_frames))
    ptd_errors = (rising_edges['stim_photodiode'][error_frames])
    photodiode_rise = np.delete(rising_edges['stim_photodiode'], error_frames)
    ptd_end -= len(error_frames)
    ptd_rise_diff = np.ediff1d(photodiode_rise)
 
# Estimate delay between stim and photodiode
 
stim_on_photodiode_idx = 60 + 120 * np.arange(0, ptd_end - ptd_start, 1)
stim_on_photodiode = falling_edges['vsync_stim'][stim_on_photodiode_idx]
photodiode_on = rising_edges['stim_photodiode'][ptd_start + np.arange(0, ptd_end - ptd_start, 1)]
delay_rise = photodiode_on - stim_on_photodiode
delay_mu = np.mean(delay_rise[:-1])
delay_sd = np.std(delay_rise[:-1])
print("monitor delay: " + str(delay_mu) + '+,-' + str(delay_sd))
 
# adjust stimulus time to incorporate monitor delay
stim_time = falling_edges['vsync_stim'] + delay_mu
 
# convert stimulus frames into twop frames
frames_2p = np.empty((len(stim_time), 1))
for i in range(len(stim_time)):
    crossings = np.searchsorted(falling_edges['vsync_2p'], stim_time[i], side='left') - 1
    if crossings < (len(falling_edges['vsync_2p']) - 1):
        frames_2p[i] = crossings
    else:
        frames_2p[i : len(stim_time)] = np.NaN
        warnings.warn('Acquisition ends before stimulus.', RuntimeWarning)
 
# Read Pickle File          
 
 
# Generate Table containing stimulus informarion of each condition:
# - Start and End 2p Frame
# - Spatial Freq
# - Temporal Freq
# - Orientation
# - Contrast
 
stimuli = stim_info['stimuli']
condition_names = ['drifting_gratings_contrast', 'drifting_gratings_TF']
 
stim_table = {}
for condition_name in condition_names:
    stim_index = -1
    for i_stim, stim_data in enumerate(stimuli):
        if condition_name in stim_data['stim_path']:
            stim_index = i_stim
 
    if stim_index > -1:
        stimulus = stimuli[stim_index]
 
        # compute display sequence
        display_sequence = np.array(stimulus['display_sequence'])
        pre_blank_sec = int(stim_info['pre_blank_sec'])
        fps = stim_info['fps']
        display_sequence += pre_blank_sec
        display_sequence *= int(stim_info['fps'])  # in stimulus frames
 
        sweep_frames = stimulus['sweep_frames']
        timing_table = pd.DataFrame(np.array(sweep_frames).astype(int), columns=('start', 'end'))
        timing_table['dif'] = timing_table['end'] - timing_table['start']
 
        timing_table.start += display_sequence[0, 0]
        for seg in range(len(display_sequence) - 1):
            for index, row in timing_table.iterrows():
                if row.start >= display_sequence[seg, 1]:
                    timing_table.start[index] = timing_table.start[index] - display_sequence[seg, 1]+ display_sequence[seg + 1, 0]
        timing_table.end = timing_table.start + timing_table.dif
 
        expected_sweeps = len(timing_table)
        timing_table = timing_table[timing_table.end <= display_sequence[-1, 1]]
        timing_table = timing_table[timing_table.start <= display_sequence[-1, 1]]
        actual_sweeps = len(timing_table)
 
        print(condition_name)
        print('Found ' + str(actual_sweeps) + ' of ' + str(expected_sweeps) + ' expected sweeps.')
 
        stim_attributes = stimulus['dimnames'] # ['Contrast', 'TF', 'SF', 'Ori']
        stim_table_condition = pd.DataFrame(np.column_stack((frames_2p[timing_table['start']],frames_2p[timing_table['end']])), columns=('Start', 'End'))
 
        sweep_table = stimulus['sweep_table']
        sweep_order = stimulus['sweep_order']
 
        num_sweeps = len(sweep_order)
        unique_conditions = np.unique(sweep_order)
 
        for i_attribure, stim_attribute in enumerate(stim_attributes):
            attribute_by_sweep = np.zeros((num_sweeps,))
            attribute_by_sweep[:] = np.NaN
            for i_condition, condition in enumerate(unique_conditions):
                sweeps_with_condition = np.argwhere(sweep_order == condition)[:, 0]
                if condition > -1:  # blank sweep is -1
                    if stim_attribute.find('Size')!=-1:
                        attribute_by_sweep[sweeps_with_condition] = sweep_table[condition][i_attribure][0]
                    else:
                        attribute_by_sweep[sweeps_with_condition] = sweep_table[condition][i_attribure]
 
            stim_table_condition[stim_attribute] = attribute_by_sweep[:len(stim_table_condition)]
 
        stim_table[condition_name] = stim_table_condition
    else:        
        print('Condition is not found for this experiment!')
 
# Adjust stim times for the MesoScope (divided by 4, should be have different start/end for each plane based on depth?)
for condition_name in stim_table:
    stim_table_condition = stim_table[condition_name]
    stim_table_condition['Start'] = np.floor(stim_table_condition.Start.values/4.0)
    stim_table_condition['End'] = np.floor(stim_table_condition.End.values/4.0)
    stim_table[condition_name] = stim_table_condition
 
 

<KeysViewHDF5 ['data', 'meta']>
ptd_start: 7
Photodiode events before stimulus start.  Deleted.
truncating photodiode to end before stim_vsync.
ptd_end: 1877 max photodiode 3778.78754 max stim 3773.22675
monitor delay: 0.0380759657570904+,-8.353538418554458e-05
drifting_gratings_contrast
Found 560 of 615 expected sweeps.
drifting_gratings_TF
Found 560 of 615 expected sweeps.


In [22]:
stim_table.keys()

dict_keys(['drifting_gratings_contrast', 'drifting_gratings_TF'])

In [30]:
stim_table['drifting_gratings_contrast']

Start      End  Contrast   TF    SF    Ori
0      243.0    261.0      0.10  1.0  0.04  315.0
1      271.0    290.0      0.80  1.0  0.04   90.0
2      300.0    318.0      0.20  1.0  0.04   90.0
3      328.0    347.0      0.10  1.0  0.04  225.0
4      357.0    376.0      0.10  1.0  0.04  315.0
..       ...      ...       ...  ...   ...    ...
555  27509.0  27527.0      0.10  1.0  0.04    0.0
556  27537.0  27556.0      0.40  1.0  0.04   45.0
557  27566.0  27585.0      0.40  1.0  0.04   45.0
558  27594.0  27613.0      0.40  1.0  0.04   90.0
559  27623.0  27642.0      0.05  1.0  0.04   90.0

[560 rows x 6 columns]

In [24]:
print(stim_table['drifting_gratings_contrast'].Contrast.unique())
print(stim_table['drifting_gratings_contrast'].SF.unique())
print(stim_table['drifting_gratings_contrast'].TF.unique())
print(stim_table['drifting_gratings_contrast'].Ori.unique())

[0.1  0.8  0.2  0.05 0.4   nan]
[0.04  nan]
[ 1. nan]
[315.  90. 225. 270.   0. 135. 180.  45.  nan]


### There are some NaN trials
- omission?

In [26]:
np.where(stim_table['drifting_gratings_contrast'].Ori.isna())

(array([ 70, 107, 134, 176, 217, 225, 239, 317, 332, 340, 386, 391, 470,
        509, 519], dtype=int64),)

In [27]:
np.where(stim_table['drifting_gratings_contrast'].Contrast.isna())

(array([ 70, 107, 134, 176, 217, 225, 239, 317, 332, 340, 386, 391, 470,
        509, 519], dtype=int64),)

In [28]:
stim_table['drifting_gratings_contrast'].iloc[70]

Start       2241.0
End         2260.0
Contrast       NaN
TF             NaN
SF             NaN
Ori            NaN
Name: 70, dtype: float64

In [8]:
pprint(list(df.keys()))

['session_uuid',
 'trigger_delay_sec',
 'syncpulselines',
 'di',
 'mouse_id',
 'nidaq_tasks',
 'post_blank_sec',
 'comp_id',
 'monitor_brightness',
 'stopdatetime',
 'user_id',
 'monitor',
 'script',
 'pre_blank_sec',
 'rig_id',
 'startdatetime',
 'platform',
 'window',
 'wwidth',
 'params',
 'stop_time',
 'scripttext',
 'droppedframes',
 'config',
 'do',
 'syncpulse',
 'monitor_contrast',
 'start_time',
 'escape_pressed',
 'wheight',
 'syncsqrsize',
 'syncpulseport',
 'vsynccount',
 'stage',
 'fps',
 'intervalsms',
 'sweepstim_text',
 'stimuli',
 'config_path',
 'movie_output',
 'sync_sqr_loc',
 'items',
 'showmouse',
 'unpickleable',
 'total_frames',
 'primary_stimulus',
 'sync_config']


## Compare this to allensdk method
- Why 560 out of 615 "expected" sweeps?

In [20]:
from allensdk.core.brain_observatory_cache import BrainObservatoryCache
from allensdk.brain_observatory.drifting_gratings import DriftingGratings

In [22]:
boc =  BrainObservatoryCache()

In [24]:
data_set = boc.get_ophys_experiment_data(502376461)

2024-06-07 11:26:03,349 allensdk.api.api.retrieve_file_over_http INFO     Downloading URL: http://api.brain-map.org/api/v2/well_known_file_download/514609062


### I need a hack for allensdk
- Omid has one branch for this, nut I need to incorporate it into Matt's and my hacks.
- Instead, I will use it directly in codeocesn for now.
    - I need a better processing pipeline to reliably calculate the response anyway.
    - Just check if his hack is the same as the one above.
